session_rpcserver.go

```diff
@@ -979,43 +1015,130 @@ func (s *sessionRpcServer) AddAutopilotSession(ctx context.Context,
+				"group %x", groupSess.ID, groupSess.GroupID)
+		}
+
+		// Now we need to check that all the sessions in the group are
+		// no longer active.
+		ok, err := s.cfg.db.CheckSessionGroupPredicate(
+			groupID, func(s *session.Session) bool {
+				return s.State == session.StateRevoked ||
+					s.State == session.StateExpired
+			},
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		if !ok {
+			return nil, fmt.Errorf("a linked session in group "+
+				"%x is still active", groupID)
+		}
+
+		linkedGroupID = &groupID
+		linkedGroupSession = groupSess
+	}
+
+	// Reserve Session Key & ID.
+	id, localPrivKey, err := s.cfg.db.ReserveNewSessionID()
+	if err != nil {
+		return nil, err
+	}
+
+	cancelIDRes := func() {
+		err = s.cfg.db.CancelIDReservation(id)
+		if err != nil {
+			log.Errorf("could not cancel reservation for "+
+				"session ID %x", id)
+		}
+	}
+
 	sess, err := session.NewSession(
-		req.Label, session.TypeAutopilot, expiry, req.MailboxServerAddr,
-		req.DevServer, perms, caveats, featureConfig, privacy,
+		id, localPrivKey, req.Label, session.TypeAutopilot, expiry,
+		req.MailboxServerAddr, req.DevServer, perms, caveats,
+		featureConfig, privacy, linkedGroupID,
 	)
 	if err != nil {
+		cancelIDRes()
+
 		return nil, fmt.Errorf("error creating new session: %v", err)
 	}
 
+	// If this session is being linked to a previous one, then we need to
+	// use the previous session's local private key to sign the new
+	// session's public key in order to prove to the Autopilot server that
+	// the two session's belong to the same owner.
+	var (
+		linkSig        []byte
+		prevSessionPub *btcec.PublicKey
+	)
+	if linkedGroupID != nil {
+		privKey := linkedGroupSession.LocalPrivateKey
+		pubKey := sess.LocalPublicKey.SerializeCompressed()
+
+		msg := chainhash.HashB(pubKey)
+		linkSig = ecdsa.Sign(privKey, msg).Serialize()
+
+		prevSessionPub = linkedGroupSession.LocalPublicKey
+	}
+
 	// Register all the privacy map pairs for this session ID.
-	privDB := s.cfg.privMap(sess.ID)
+	privDB := s.cfg.privMap(sess.GroupID)
 	err = privDB.Update(func(tx firewalldb.PrivacyMapTx) error {
 		for r, p := range privacyMapPairs {
 			err := tx.NewPair(r, p)
 			if err != nil {
+				cancelIDRes()
+
 				return err
 			}
 		}
 		return nil
 	})
 	if err != nil {
+		cancelIDRes()
+
 		return nil, err
 	}
 
 	// Attempt to register the session with the Autopilot server.
 	remoteKey, err := s.cfg.autopilot.RegisterSession(
 		ctx, sess.LocalPublicKey, sess.ServerAddr, sess.DevServer,
-		featureConfig,
+		featureConfig, prevSessionPub, linkSig,
 	)
 	if err != nil {
+		cancelIDRes()
+
 		return nil, fmt.Errorf("error registering session with "+
 			"autopilot server: %v", err)
 	}
 
 	// We only persist this session if we successfully retrieved the
 	// autopilot's static key.
 	sess.RemotePublicKey = remoteKey
-	if err := s.db.StoreSession(sess); err != nil {
+	if err := s.cfg.db.CreateSession(sess); err != nil {
```

Comment:
Same here, should we add `cancelIDRes()`?

Reply:
updated!